package NYT.NNodeTrackerClient.NProto;

import "yt/core/misc/guid.proto";
import "yt/core/misc/error.proto";
import "yt/ytlib/node_tracker_client/node.proto";
import "yt/ytlib/chunk_client/chunk_meta.proto";
import "yt/ytlib/chunk_client/chunk_info.proto";
import "yt/ytlib/hydra/hydra_manager.proto";

////////////////////////////////////////////////////////////////////////////////

// Chunk Manager stuff.

message TChunkAddInfo
{
    required NYT.NProto.TGuid chunk_id = 1;
    required bool cached = 2;
    required bool active = 3;
    required NYT.NChunkClient.NProto.TChunkInfo chunk_info = 4;
}

message TChunkRemoveInfo
{
    required NYT.NProto.TGuid chunk_id = 1;
    required bool cached = 2;
}

////////////////////////////////////////////////////////////////////////////////

message TReqRegisterNode
{
    required TNodeDescriptor node_descriptor = 1;
    required TNodeStatistics statistics = 2;
}

message TRspRegisterNode
{
    required int32 node_id = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqFullHeartbeat
{
    required int32 node_id = 1;
    required TNodeStatistics statistics = 2;

    // Chunk Manager stuff.
    repeated TChunkAddInfo chunks = 3;
}

message TRspFullHeartbeat
{ }

////////////////////////////////////////////////////////////////////////////////

message TTabletSlotInfo
{
    required int32 peer_state = 1;
    optional NYT.NProto.TGuid cell_id = 2;
    optional int32 peer_id = 3;
    optional int32 config_version = 4;
    optional NYT.NProto.TGuid prerequisite_transaction_id = 5;
}

message THiveCellInfo
{
    required NYT.NProto.TGuid cell_id = 1;
    required int32 config_version = 2;
}

message TTabletStatistics
{
    required int32 partition_count = 1;
    required int32 store_count = 2;
}

message TTabletInfo
{
    required NYT.NProto.TGuid tablet_id = 1;
    required TTabletStatistics statistics = 2;
}

message TReqIncrementalHeartbeat
{
    required int32 node_id = 1;

    // Node Tracker stuff.
    required TNodeStatistics statistics = 2;
    repeated string alerts = 3;

    // Chunk Manager stuff.
    repeated TChunkAddInfo added_chunks = 4;
    repeated TChunkRemoveInfo removed_chunks = 5;

    // Tablet Manager stuff.
    repeated TTabletSlotInfo tablet_slots = 6;
    repeated TTabletInfo tablets = 8;

    // Hive Manager stuff.
    repeated THiveCellInfo hive_cells = 7;
}

message TCreateTabletSlotInfo
{
    required NYT.NProto.TGuid cell_id = 1;
    required bytes options = 2; // NTabletClient::TTabletCellOptions
    required NYT.NProto.TGuid prerequisite_transaction_id = 3;
}

message TConfigureTabletSlotInfo
{
    required NYT.NProto.TGuid cell_id = 1;
    required int32 config_version = 2;
    required bytes config = 3; // NTabletClient::TTabletCellConfig
    required int32 peer_id = 4;
}

message TRemoveTabletSlotInfo
{
    required NYT.NProto.TGuid cell_id = 1;
}

message TUnregisterHiveCellInfo
{
    required NYT.NProto.TGuid cell_id = 1;
}

message TReconfigureHiveCellInfo
{
    required int32 config_version = 1;
    required bytes config = 2; // NElection::TCellConfig
}

message TRspIncrementalHeartbeat
{
    // Tablet Manager stuff.
    repeated TCreateTabletSlotInfo tablet_slots_to_create = 1;
    repeated TRemoveTabletSlotInfo tablet_slots_to_remove = 2;
    repeated TConfigureTabletSlotInfo tablet_slots_configure = 3;

    // Hive Manager stuff.
    repeated TUnregisterHiveCellInfo hive_cells_to_unregister = 4;
    repeated TReconfigureHiveCellInfo hive_cells_to_reconfigure = 5;
}

////////////////////////////////////////////////////////////////////////////////
